contracts/examples/lottery-esdt/src/lottery.rs

```diff
@@ -277,33 +274,161 @@ pub trait Lottery {
         } else {
             info.prize_distribution.len()
         };
-        let total_prize = info.prize_pool.clone();
-        let winning_tickets = self.get_distinct_random(1, total_tickets, total_winning_tickets);
-
-        // distribute to the first place last. Laws of probability say that order doesn't matter.
-        // this is done to mitigate the effects of BigUint division leading to "spare" prize money being left out at times
-        // 1st place will get the spare money instead.
-        for i in (1..total_winning_tickets).rev() {
-            let winning_ticket_id = winning_tickets[i];
-            let winner_address = ticket_holders_mapper.get(winning_ticket_id);
-            let prize = self.calculate_percentage_of(
-                &total_prize,
-                &BigUint::from(info.prize_distribution.get(i)),
-            );
 
+        self.total_winning_tickets(lottery_name)
+            .set(total_winning_tickets);
+        self.index_last_winner(lottery_name).set(1);
+    }
+
+    fn distribute_prizes(&self, lottery_name: &ManagedBuffer) -> AwardingStatus {
+        let mut info = self.lottery_info(lottery_name).get();
+        let ticket_holders_mapper = self.ticket_holders(lottery_name);
+        let total_tickets = ticket_holders_mapper.len();
+
+        let mut index_last_winner = self.index_last_winner(lottery_name).get();
+        let total_winning_tickets = self.total_winning_tickets(lottery_name).get();
+        require!(
+            index_last_winner <= total_winning_tickets,
+            "Awarding has ended"
+        );
+
+        let mut iterations = 0;
+        while index_last_winner <= total_winning_tickets && iterations < MAX_OPERATIONS {
+            let rand_index = self.get_distinct_random(index_last_winner, total_tickets, 1)[0];
+
+            // swap indexes of the winner addresses - we are basically bringing the winners in the first indexes of the mapper
+            let winner_address = self.ticket_holders(lottery_name).get(rand_index).clone();
+            let last_index_winner_address =
+                self.ticket_holders(lottery_name).get(index_last_winner);
+            self.ticket_holders(lottery_name)
+                .set(rand_index, &last_index_winner_address);
+            self.ticket_holders(lottery_name)
+                .set(index_last_winner, &winner_address);
+
+            // distribute to the first place last. Laws of probability say that order doesn't matter.
+            // this is done to mitigate the effects of BigUint division leading to "spare" prize money being left out at times
+            // 1st place will get the spare money instead.
+            if index_last_winner < total_winning_tickets {
+                let prize = self.calculate_percentage_of(
+                    &info.prize_pool,
+                    &BigUint::from(info.prize_distribution.get(index_last_winner)),
```

Comment:
I think it should be this way, since you say you'll do the distribution for the first place last.

```suggestion
                    &BigUint::from(info.prize_distribution.get(index_last_winner + 1)),
```

Reply:
Nvm, that's a ManagedVec, starts from 0.